### PR TITLE
Modified get method to search by filters

### PR DESCRIPTION
**Pull request by Bruno-Israel-Mendoza-Aguilar**
Implemented changes on controller,service,repository layers and created exception for unknown parameter names.

### CODE REVIEW

File: src/main/resources/application.properties [hunk 1]
```diff
@@ -1,4 +1,4 @@
 server.port=5000
-spring.data.mongodb.uri=mongodb+srv://db_user:[password]@cluster0-nomfh.mongodb.net/InternHome?retryWrites=true&w=majority
+spring.data.mongodb.uri=mongodb+srv://db_user:vloUr7AakeTEdXDj@cluster0-nomfh.mongodb.net/InternHome?retryWrites=true&w=majority
```
Comment (VictorMunguiaEnciso) [inline src/main/resources/application.properties]: Please, delete the password

File: src/main/java/com/sprints/swagger/SprintsSwagger.java [hunk 2]
```diff
@@ -1,23 +1,23 @@
...
-	@Bean
-	public Docket api() {
-		return new Docket(DocumentationType.SWAGGER_2)
-				.select()
-				.apis(RequestHandlerSelectors.any())
-				.paths(PathSelectors.any())
-				.build();
-	}
-}
+package com.sprints.swagger;
```
Comment (VictorMunguiaEnciso) [inline src/main/java/com/sprints/swagger/SprintsSwagger.java]: Why did you modified the swagger class ?

File: src/main/java/com/sprints/repository/SprintsCustomRepositoryImpl.java [hunk 3]
```diff
@@ -28,5 +34,32 @@ public Sprint oneSprintBacklogValidation() {
...
-	
-}
+
+	@Override
+	public List<Sprint> findAllByParams(Optional<String> name, Optional<String> technology,
+			Optional<LocalDate> start_date, Optional<LocalDate> end_date) {
+		
+		Criteria criteria = new Criteria();
+		
+		if(name.isPresent()) {
```
Comment (VictorMunguiaEnciso) [inline src/main/java/com/sprints/repository/SprintsCustomRepositoryImpl.java]: In the repository we should not have any if or validation. Le´st try to move all the IFs outside of the repository,

File: src/main/java/com/sprints/interceptors/UndeclaredParamsHandlerInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;
+
+import com.sprints.exception.UndeclaredRequestParamException;
+
+import org.springframework.util.StringUtils;
+
+
+
+public class UndeclaredParamsHandlerInterceptor extends HandlerInterceptorAdapter{
```
Comment (VictorMunguiaEnciso) [inline src/main/java/com/sprints/interceptors/UndeclaredParamsHandlerInterceptor.java]: What is the objective of this class ?

File: src/main/java/com/sprints/interceptors/DisallowUndeclaredRequestParams.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.sprints.interceptors;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface DisallowUndeclaredRequestParams {
```
Comment (VictorMunguiaEnciso) [inline src/main/java/com/sprints/interceptors/DisallowUndeclaredRequestParams.java]: What is the objective of this class  ?

File: src/main/java/com/sprints/controllers/SprintsController.java [hunk 6]
```diff
@@ -48,8 +53,18 @@ public SprintDomain findSprintById(@PathVariable("id") String id) throws EntityN
...
 	@RequestMapping(method=RequestMethod.GET)
-	public List<SprintDomain> findAllSprints(){
-		return sprintsService.findAll();
+	@DisallowUndeclaredRequestParams
+	public List<SprintDomain> findAllSprints( @RequestParam("name") Optional<String> name,
+												@RequestParam("technology") Optional<String> technology,
+												@RequestParam("start_date") @DateTimeFormat(iso = ISO.DATE) Optional<LocalDate> start_date,
+												@RequestParam("end_date") @DateTimeFormat(iso = ISO.DATE) Optional<LocalDate> end_date){
+		
+		if(name.isEmpty() && technology.isEmpty() && start_date.isEmpty() && end_date.isEmpty()) {
```
Comment (VictorMunguiaEnciso) [inline src/main/java/com/sprints/controllers/SprintsController.java]: We well need take this if to serviceImpl

File: src/main/java/com/sprints/SprintsApiApplication.java [hunk 7]
```diff
@@ -3,12 +3,27 @@
...
 @SpringBootApplication(exclude = {DataSourceAutoConfiguration.class })
 public class SprintsApiApplication {
 	public static void main(String[] args) {
 		SpringApplication.run(SprintsApiApplication.class, args);
 		    
-		
+	}
+	
+	@Configuration
+	public static class WebMvcConfig extends WebMvcConfigurerAdapter  {
```
Comment (VictorMunguiaEnciso) [inline src/main/java/com/sprints/SprintsApiApplication.java]: Why we need this method ?

File: src/main/java/com/sprints/SprintsApiApplication.java [hunk 8]
```diff
@@ -3,12 +3,27 @@
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.web.servlet.config.annotation.InterceptorRegistry;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter;
 
+import com.sprints.interceptors.UndeclaredParamsHandlerInterceptor;
+
+@SuppressWarnings("deprecation")
```
Comment (VictorMunguiaEnciso) [inline src/main/java/com/sprints/SprintsApiApplication.java]: Why we need suppress deprecate warnings ? We should not use deprecate implementations

File: src/main/java/com/sprints/SprintsApiApplication.java (hunk 7 above)
Comment (Bruno-Israel-Mendoza-Aguilar) [reply src/main/java/com/sprints/SprintsApiApplication.java]: That class is needed as configuration so I can register the Interceptors for the undeclared params in the Get filters

File: src/main/java/com/sprints/interceptors/DisallowUndeclaredRequestParams.java (hunk 5 above)
Comment (Bruno-Israel-Mendoza-Aguilar) [reply src/main/java/com/sprints/interceptors/DisallowUndeclaredRequestParams.java]: It´s the interface I used for the Interceptor for undeclared params on Get operation filters

File: src/main/java/com/sprints/interceptors/UndeclaredParamsHandlerInterceptor.java (hunk 4 above)
Comment (Bruno-Israel-Mendoza-Aguilar) [reply src/main/java/com/sprints/interceptors/UndeclaredParamsHandlerInterceptor.java]: There I implement methods like preHandle to get the parameters sent in the Get operation filters by the user and validate them before they get to the controller. Also it contains the method to get the names of valid params and compares it with the values sent by the user and throws an exception if they´re invalid

File: src/main/java/com/sprints/swagger/SprintsSwagger.java (hunk 2 above)
Comment (Bruno-Israel-Mendoza-Aguilar) [reply src/main/java/com/sprints/swagger/SprintsSwagger.java]: Sorry about this, I don´t really remember touching anything from this file, I checked and for some reason Git shows it like I replaced the whole file with the same information that was already there before. I don´t know why that happened

File: src/main/java/com/sprints/SprintsApiApplication.java (hunk 8 above)
Comment (Bruno-Israel-Mendoza-Aguilar) [reply src/main/java/com/sprints/SprintsApiApplication.java]: Fixed

File: src/main/java/com/sprints/controllers/SprintsController.java (hunk 6 above)
Comment (Bruno-Israel-Mendoza-Aguilar) [reply src/main/java/com/sprints/controllers/SprintsController.java]: Fixed

File: src/main/java/com/sprints/repository/SprintsCustomRepositoryImpl.java (hunk 3 above)
Comment (Bruno-Israel-Mendoza-Aguilar) [reply src/main/java/com/sprints/repository/SprintsCustomRepositoryImpl.java]: Fixed

File: src/main/resources/application.properties (hunk 1 above)
Comment (Bruno-Israel-Mendoza-Aguilar) [reply src/main/resources/application.properties]: Fixed

**Pull request merged**